### ISSUE TITLE
AppId Issue

### ISSUE THREAD

**Issue by WatameBytes** [Yubico/java-webauthn-server#389]
**tl;dr: I didn't set an AppId, and it's causing me issues at the moment. I'm thinking of making a custom deserializer.** 

I'm aware AppId is optional and not required. This is a new test application I'm working on, so I doubt I have any credentials registered via U2F. 

The problem is, when I don't set an AppId, causes problems that I've been mostly been able to resolve and do have a potential solution for this problem, but wanted to see if there was an easier way.

Going to give some context, skip to the Assertion section if you want to get to what I want an answer to.

Register - This is what I get back from the first endpoint
```
{
    "publicKeyCredentialCreationOptions": {
        "rp": {
            "name": "Example Application",
            "id": "localhost"
        },
        "user": {
          // User info
        },
        "challenge": "A CHALLENGE",
        "pubKeyCredParams": [
            // List of algos
        ],
        "timeout": 60000,
        "excludeCredentials": [],
        "authenticatorSelection": {
            // info
        },
        "attestation": "none",
        "extensions": { <--- This field is giving me problems
            "appidExclude": null,
            "credProps": true,
            "largeBlob": null
        }
    }
}
```
Extensions having null has been giving me some problems in the frontend, so **I used MixIns to remove the extension part**. 

Which works fine for this. 

We have a database that stores this response and when we retrieve it we use `return objectMapper.readValue(json, clazz);` to deserialize it back into `PublicKeyCredentialCreationOptions.class`. 

Assertion
This is what my assertion request object looks like, before I apply any MixIns
```
{
    "publicKeyCredentialRequestOptions": {
        "challenge": "CHALLENGE",
        "timeout": 60000,
        "rpId": "localhost",
        "userVerification": "preferred",
        **"extensions": { <-- This is causing problems
            "appid": null
        }**
    }
}
```

I used a MixIn to convert it into 
```
{
    "assertionId": "2ee4f1be-5d60-4081-8fa9-3cd67f21ad1c",
    "publicKeyCredentialRequestOptions": {
        "challenge": "gAd4DgQ1xpEQGfwGvomnZTgxbh7KuRKtCL1JdhjsWi0",
        "timeout": 60000,
        "rpId": "localhost",
        "userVerification": "preferred"
    }
}
```
Mostly because my frontend code doesn't like it when appid is null.

Here is the problem I am facing.

When I attempt to deserialize that object, I keep getting `Caused by: java.lang.NullPointerException: extensions is marked non-null but is null`. 

I was thinking of making a custom deserializer to address this problem, but wanted to see if there was an easier way of achieving this.

I believe the issue is with

In `PublicKeyCredentialCreationOptions.class`, we don't check if Extensions is null. 
```
    @JsonCreator
    private PublicKeyCredentialCreationOptions(@JsonProperty("rp") @NonNull RelyingPartyIdentity rp, @JsonProperty("user") @NonNull UserIdentity user, @JsonProperty("challenge") @NonNull ByteArray challenge, @JsonProperty("pubKeyCredParams") @NonNull List<PublicKeyCredentialParameters> pubKeyCredParams, @JsonProperty("timeout") Long timeout, @JsonProperty("excludeCredentials") Set<PublicKeyCredentialDescriptor> excludeCredentials, @JsonProperty("authenticatorSelection") AuthenticatorSelectionCriteria authenticatorSelection, @JsonProperty("attestation") AttestationConveyancePreference attestation, @JsonProperty("extensions") RegistrationExtensionInputs extensions) {
        if (rp == null) {
            throw new NullPointerException("rp is marked non-null but is null");
        } else if (user == null) {
            throw new NullPointerException("user is marked non-null but is null");
        } else if (challenge == null) {
            throw new NullPointerException("challenge is marked non-null but is null");
        } else if (pubKeyCredParams == null) {
            throw new NullPointerException("pubKeyCredParams is marked non-null but is null");
        } else {
            this.rp = rp;
            this.user = user;
            this.challenge = challenge;
            this.pubKeyCredParams = filterAvailableAlgorithms(pubKeyCredParams);
            this.timeout = timeout;
            this.excludeCredentials = excludeCredentials == null ? null : CollectionUtil.immutableSortedSet(new TreeSet(excludeCredentials));
            this.authenticatorSelection = authenticatorSelection;
            this.attestation = attestation == null ? AttestationConveyancePreference.NONE : attestation;
            this.extensions = extensions == null ? RegistrationExtensionInputs.builder().build() : extensions;
        }
    }
```

In Assertion, or more specially `PublicKeyCredentialRequestOptions.class`, we do check if extensions is null and throw an error versus us having `RegistrationExtensionInputs.builder().build()`.
```
    @JsonCreator
    private PublicKeyCredentialRequestOptions(@JsonProperty("challenge") @NonNull ByteArray challenge, @JsonProperty("timeout") Long timeout, @JsonProperty("rpId") String rpId, @JsonProperty("allowCredentials") List<PublicKeyCredentialDescriptor> allowCredentials, @JsonProperty("userVerification") UserVerificationRequirement userVerification, @JsonProperty("extensions") @NonNull AssertionExtensionInputs extensions) {
        if (challenge == null) {
            throw new NullPointerException("challenge is marked non-null but is null");
        } else if (extensions == null) {
            throw new NullPointerException("extensions is marked non-null but is null");
        } else {
            this.challenge = challenge;
            this.timeout = timeout;
            this.rpId = rpId;
            this.allowCredentials = allowCredentials == null ? null : CollectionUtil.immutableList(allowCredentials);
            this.userVerification = userVerification;
            this.extensions = extensions;
        }
    }
``` 

In my code

This is what my frontend gives back
```
Extension results: ClientAssertionExtensionOutputs(appid=Optional.empty, largeBlob=Optional.empty)
```

and here is my json that we have stored to compare
```
{"publicKeyCredentialRequestOptions":{"challenge":"CHALLENGE","timeout":60000,"rpId":"localhost","userVerification":"preferred"},"username":null,"userHandle":null}
```

**Comment by WatameBytes** [comment]
It did work when I set an AppId to "https://localhost:3000", but now I'm running an application that can't run in HTTPS in local development and AppId doesn't allow HTTP.

**Comment by emlun** [comment]
Hi, have you tried using the `toCredentialsCreateJson` and `parseRegistrationResponseJson` methods as explained in the [getting started guide](https://github.com/Yubico/java-webauthn-server?tab=readme-ov-file#getting-started)?

**Comment by WatameBytes** [comment]
Thank you, that was the solution to that problem.

At the moment, not sure if there is a way to prioritize Chrome Passkeys over Windows Security, I did try modifying the selection, but no result. 

Not sure if you happen to know the solution to that, if not it's fine. The main issue has been resolved. Thank you.

**Comment by emlun** [comment]
I don't think there is. The closest thing is the new [`hints` parameter](https://w3c.github.io/webauthn/#dom-publickeycredentialcreationoptions-hints) coming in WebAuthn L3 which almost does this, but it only prioritizes the _kind_ of authenticator and attachment, not the provider within one of those categories.
